write-fonts/src/tables/cmap.rs

```diff
@@ -238,6 +237,222 @@ impl Cmap {
+                if self.gids_in_order {
+                    return Some(self.make_segment(i));
+                }
+            // and the funny case:
+            // if gids were not previously ordered but are now:
+            // - if i == 0, then this is the first item in a new segment;
+            //   set gids_in_order and continue
+            // - if i > 0, we need to back up one
+            } else if !self.gids_in_order {
+                if i == 0 {
+                    self.gids_in_order = true;
+                } else {
+                    return Some(self.make_segment(i - 1));
+                }
+            }
+            prev_cp = *cp;
+            prev_gid = *gid;
+        }
+
+        // if we're done looping then create the last segment:
+        let last_idx = self.mappings.len() - 1;
+        Some(self.make_segment(last_idx - self.seg_start))
+    }
+
+    /// Compute an efficient set of segments.
+    ///
+    /// - A segment is a contiguous range of chars.
+    /// - If all the chars in a segment share a common delta to their glyph ids,
+    ///   we can encode them much more efficiently
+    /// - it's possible for a contiguous range of chars to contain a subrange
+    ///   that share a common delta, where the overall range does not, e.g.
+    ///
+    ///   ```text
+    ///   [a b c d e f g]
+    ///   [9 3 6 7 8 2 1]
+    ///   ```
+    ///   (here a-g is a range containing the subrange c-e, which have a common
+    ///   delta.)
+    ///
+    /// This leads us to a reasonably intuitive algorithm: we start by greedily
+    /// splitting ranges up so we can consider all subranges with common deltas;
+    /// then we look at these one at a time, and combine them back together if
+    /// doing so saves space.
+    ///
+    /// This differs from the python, which starts from larger segments and then
+    /// subdivides them, but the overall idea is the same.
+    ///
+    /// <https://github.com/fonttools/fonttools/blob/f1d3e116d54f/Lib/fontTools/ttLib/tables/_c_m_a_p.py#L783>
+    fn compute(mut self) -> Vec<Format4Segment> {
+        let Some(first) = self.next_possible_segment() else {
+            return Default::default();
+        };
+
+        let mut result = vec![first];
+
+        // now we want to collect the segments, combining smaller segments where
+        // that leads to a size savings.
+        let mut next = self.next_possible_segment();
+
+        while let Some(current) = next.take() {
+            next = self.next_possible_segment();
+            let prev = result.last_mut().unwrap();
+
+            let can_merge_left = prev.can_merge(&current);
+
+            if !can_merge_left {
+                result.push(current);
+                continue;
+            }
+
+            let combined = prev.combine(current);
+            let should_merge = match next.as_ref() {
+                // if we can merge on both sides, consider total cost
```

Comment:
Why do you need to check both sides at once? That is, why doesn't checking if current could merge with the last entry in result suffice?

Reply:
Per IM, let's add a comment explaining the arithmatic of merging that leads to potential for merging 3 to be a win where 2 wasn't, and no need to check 4+